### PR TITLE
Source walk/Spring clean/Parameterize/Enhance/Squash warnings et.al.

### PR DESCRIPTION
**Pull request by nickl-**
Note: This was also based from #157 my apologies in advance but man junit 3 is a pain when you are used to more... did make an effort to just patch support but it turned into too much effort and since it has already been done I just gave up for the easy route, not to get too distracted from the task I've set ahead.

## Goals
 * Do detailed code review and maintenance
     * refactoring for optimised readability/usability
     * modernise for 1.6 source dropping legacy and redundancy
     * identify smells
 * Get compiler friendly
     * Squash all compiler warnings
     * Parameterise types to benefit from conversion assistance
     * Apply `java.lang` annotations for optimal compiler support.
 
## Accomplished
Managed to do 3 repetitive code sweeps after first clearing the deprecation warnings. All those fixes, yes fixed not suppressed, are committed separately. First sweep applied parameters, then came back to take advantage of the new types by enhancing loops and removing unnecessary casts, applied the annotations and went through for another pass to verify while removing redundant else statements. Reviewed every change again before committing applying the smallest style fixes without getting carried away. After commit fixed up unit tests for compiler warnings and ran import organisation as the final polish to dissolve any ambiguity.

**please note: only annotations and imports were automated the rest was done by hand.**
**please note: this patch excludes files from my other PRs so as not to cause merge conflicts.**

### More details about changes
 * Parameterize raw types.

All raw types were given parameters and no unchecked conversions were required (there is one unchecked conversion in the `scopedpool.SoftValueHashMap` as a result of the way the reference queue is implemented which is unavoidable and got suppressed but that is not included in this patch see #159) There are 2 occurrences where generic types could not be utilised and had to revert to general Object usage, these occurrences were noted a smells and can be revisited to be rectified and a subsequent task.

 * Type widening

Applied type widening on all collection references to the base interfaces List, Map or Set as appropriate. This of course will have zero impact for compatibility where parameters are concerned. Where specific types were returned in the past, will have a limited impact which depends on the dimensions of the receiving types, but easily rectified imho. The only concern is with members extending the api and overwriting these methods, previously forced to implement the narrowed types which now has signature changed so not compatible anymore. I do believe these users would still welcome the improvement which they in turn also benefit from. 

If zero compatibility impact is a goal then we can overload the change and deprecate the faulty implementation, this way the children will still work as well as the callers. Shame I didn't think of this before, I will hunt down the public methods and commit a change shortly.

All specific Reference (Weak/Soft/etc) types were also widened to use the abstract Reference base class but I don't recall these being exposed.

* hashCode implementations

There were a few custom `equals` implementations that lacked their required `hashCode` counterparts as per the `Object.equals` contract which I implemented and committed each separately.

 * Convert loops to enhance for loop.

Since enhanced for loops date back to Java 1.5 it is high time we get rid of the manual iterator implementations and counter based for loop implementations in favour of the short hand. In some places where the Iterator implementation went far out it was noted as a smell and can be revisited for modification at a later stage. 

 * Remove redundant else statements.

Not as a blanket rule but where it improved readability the redundant else was removed and the implicit then implementation moved outside the block for fall through processing. There were only 1 occurrence where the else statement was completely redundant, variable starts off as null and gets set to null again on fall through, which was removed but the rest all retained 100 % of the previous functionality.  I did not note any of the remaining unmodified implementations as smells as I consider them, does as advertised. 

 * Annotate override and deprecated methods.

With the aid of eclipse all `@Override` and `@Deprecated` annotations were applied and the compiler will henceforth assist with keeping these arrangements. The new `@Deprecated` annotation include properties `since` and `forRemoval` which I do not have the data for. These can be collected as smells and populated at a later stage.

 * Mark unused members.

Unused members ie, fields, variables and private methods were suppression annotated so that we can easily find them later and deal with each occurrence individually to decide whether it should be removed or if the purpose was intended but not completely implemented as yet.

 * Code smells.

This PR sweeps some of the dust out and lays a polished foundation but is not intended as a refactoring exorcize. Aside from the `hashCode` methods mentioned before, no functionality was changed or added with this update. The output remains the same or iow no additional unit tests are required. Whenever more complex problem areas were discovered they were flagged and passed over. With adequate spare time I would log these as issues where they may be discussed, designed, planned out, a difficulty assigned to by the community and perhaps interested parties can jump in and get them resolved. 

Called smells cause something aint smell right but hey, perhaps I am mistaken and that's what can be discussed at least the cluttered drawers are pulled out, placed under the spotlight and tidied again, every once in a while.

## Result

```console
$ mvn clean install
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building Javassist 3.22.0-GA
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ javassist ---
[INFO] Deleting /home/user/dev/javassist/target
[INFO]
[INFO] --- maven-resources-plugin:3.0.2:resources (default-resources) @ javassist ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/user/dev/javassist/src/main/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.2:compile (default-compile) @ javassist ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 218 source files to /home/user/dev/javassist/target/classes
[INFO]
[INFO] --- maven-bundle-plugin:3.3.0:manifest (bundle-manifest) @ javassist ---
[INFO]
[INFO] --- maven-resources-plugin:3.0.2:testResources (default-testResources) @ javassist ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/user/dev/javassist/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.2:testCompile (default-testCompile) @ javassist ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 267 source files to /home/user/dev/javassist/target/test-classes
[INFO]
[INFO] --- maven-surefire-plugin:2.18.1:test (default-test) @ javassist ---
[INFO] Surefire report directory: /home/user/dev/javassist/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
....
Tests run: 397, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 43.749 sec - in javassist.JvstTest

Results :

Tests run: 397, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] --- maven-bundle-plugin:3.3.0:bundle (default-bundle) @ javassist ---
[INFO]
[INFO] >>> maven-source-plugin:2.0.4:jar (attach-sources) > generate-sources @ javassist >>>
[INFO]
[INFO] <<< maven-source-plugin:2.0.4:jar (attach-sources) < generate-sources @ javassist <<<
[INFO]
[INFO]
[INFO] --- maven-source-plugin:2.0.4:jar (attach-sources) @ javassist ---
[INFO] Building jar: /home/user/dev/javassist/target/javassist-3.22.0-GA-sources.jar
[INFO]
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ javassist ---
[INFO] Installing /home/user/dev/javassist/target/javassist-3.22.0-GA.jar to /home/user/.m2/repository/org/javassist/javassist/3.22.0-GA/javassist-3.22.0-GA.jar
[INFO] Installing /home/user/dev/javassist/pom.xml to /home/user/.m2/repository/org/javassist/javassist/3.22.0-GA/javassist-3.22.0-GA.pom
[INFO] Installing /home/user/dev/javassist/target/javassist-3.22.0-GA-sources.jar to /home/user/.m2/repository/org/javassist/javassist/3.22.0-GA/javassist-3.22.0-GA-sources.jar
[INFO]
[INFO] --- maven-bundle-plugin:3.3.0:install (default-install) @ javassist ---
[INFO] Installing org/javassist/javassist/3.22.0-GA/javassist-3.22.0-GA.jar
[INFO] Writing OBR metadata
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 54.645 s
[INFO] Finished at: 2017-10-30T21:33:52+02:00
[INFO] Final Memory: 15M/52M
[INFO] ------------------------------------------------------------------------

```

nice =)

I know this is a huge patch and maybe you don't agree with everything I propose... tell me what you think and if there's anything that you'd rather want.

### CODE REVIEW

**Comment by nickl-** [comment]
Now I can't find a single public method affected... is it possible that it is my good luck? =)

**Comment by nickl-** [comment]
Not sure how long I should wait for a response on these patches from the javassist team, not kidding now I would've thought a week was a reasonable waiting time. Uptake time increases as time passes I had hoped we could've at least address some questions or concerns by now, but anyhow. It is obvious this project is in desperate need of help but as Confucius says: you can only bring donkey to water, you cannot make it drink. I need to close this chapter and move on with my life. 

There were still the recorded smells which ideally got revealed as projects, milestones or at least separate issues to rally community assistance and see them addressed. As for now I just want them off my plate and not see them devoured by the Langoliers. Since these are dependent on this patch, with all the bigger fish fried, it seems appropriate to record them here for reference but absent further explanation, we'll just have to spend the time rediscovering the reasons why.

It actually seems like I've already lost some of my notes, which is just too bad. These are the smells I managed to salvage....

HotSwapper reload and then reload2 with msg is weird

AttributeInfo should support subtypes especially tag

FieldInfo/MethodInfo could do with a super type perhaps

Iterator
src/main/javassist/ClassPool.java
src/main/javassist/CodeConverter.java
src/main/javassist/CtBehavior.java
src/main/javassist/CtClassType.java
src/main/javassist/CtConstructor.java
src/main/javassist/CtMethod.java
src/main/javassist/bytecode/ClassFile.java
src/main/javassist/bytecode/CodeAnalyzer.java
src/main/javassist/bytecode/CodeAttribute.java
src/main/javassist/bytecode/CodeIterator.java
src/main/javassist/bytecode/Descriptor.java
src/main/javassist/bytecode/InstructionPrinter.java
src/main/javassist/bytecode/MethodInfo.java
src/main/javassist/bytecode/StackMap.java
src/main/javassist/bytecode/StackMapTable.java
src/main/javassist/bytecode/analysis/Analyzer.java
src/main/javassist/bytecode/analysis/Executor.java
src/main/javassist/bytecode/analysis/FramePrinter.java
src/main/javassist/bytecode/analysis/MultiType.java
src/main/javassist/bytecode/analysis/SubroutineScanner.java
src/main/javassist/bytecode/analysis/Type.java
src/main/javassist/bytecode/analysis/Util.java
src/main/javassist/bytecode/stackmap/BasicBlock.java
src/main/javassist/bytecode/stackmap/TypeData.java
src/main/javassist/compiler/MemberResolver.java
src/main/javassist/convert/TransformAccessArrayField.java
src/main/javassist/convert/TransformAfter.java
src/main/javassist/convert/TransformBefore.java
src/main/javassist/convert/TransformCall.java
src/main/javassist/convert/TransformFieldAccess.java
src/main/javassist/convert/TransformNew.java
src/main/javassist/convert/TransformNewClass.java
src/main/javassist/convert/TransformReadField.java
src/main/javassist/convert/TransformWriteField.java
src/main/javassist/convert/Transformer.java
src/main/javassist/expr/Cast.java
src/main/javassist/expr/ConstructorCall.java
src/main/javassist/expr/Expr.java
src/main/javassist/expr/ExprEditor.java
src/main/javassist/expr/FieldAccess.java
src/main/javassist/expr/Handler.java
src/main/javassist/expr/Instanceof.java
src/main/javassist/expr/MethodCall.java
src/main/javassist/expr/NewArray.java
src/main/javassist/expr/NewExpr.java
src/main/javassist/scopedpool/ScopedClassPoolRepositoryImpl.java
src/main/javassist/util/HotSwapper.java
src/main/javassist/util/proxy/ProxyFactory.java

Generic generics 
src/main/javassist/compiler/AccessorMaker.java
src/main/javassist/CtClassType.java

Unused
src/main/javassist/bytecode/CodeAttribute.java
src/main/javassist/bytecode/SignatureAttribute.java
src/main/javassist/bytecode/stackmap/TypeData.java
src/main/javassist/compiler/JvstCodeGen.java
src/main/javassist/compiler/Lex.java
src/main/javassist/compiler/MemberCodeGen.java
src/main/javassist/compiler/Parser.java
src/main/javassist/compiler/TypeChecker.java
src/main/javassist/expr/Cast.java
src/main/javassist/expr/Handler.java
src/main/javassist/expr/Instanceof.java
src/main/javassist/util/HotSwapper.java

Deprecated
src/main/javassist/CtClass.java
src/main/javassist/CtClassType.java
src/main/javassist/bytecode/AttributeInfo.java
src/main/javassist/bytecode/DeprecatedAttribute.java
src/main/javassist/bytecode/LocalVariableAttribute.java
src/main/javassist/scopedpool/ScopedClassPool.java
src/main/javassist/util/proxy/ProxyFactory.java
src/main/javassist/util/proxy/RuntimeSupport.java

I think that's that... lets hope it still makes sense 2 years from now.

**Comment by chibash** [comment]
Hi,

Thank you for your great efforts!  Since the patches are huge, please don't hurry.  Reviewing them would take time.

Keeping this old code style is somewhat on purpose for supporting J2ME etc.  But yes, it would be time now to clean up and modernize the code.

File: src/main/javassist/bytecode/AttributeInfo.java [hunk 1]
```diff
@@ -79,65 +80,58 @@ static AttributeInfo read(ConstPool cp, DataInputStream in)
...
-                    return new SourceFileAttribute(cp, name, in);
-                else if (nameStr.equals(SyntheticAttribute.tag))
-                    return new SyntheticAttribute(cp, name, in);
-                else if (nameStr.equals(StackMap.tag))
-                    return new StackMap(cp, name, in);
-                else if (nameStr.equals(StackMapTable.tag))
-                    return new StackMapTable(cp, name, in);
-            }
-        }
+        if (first < 'E')
```
Comment (chibash) [inline src/main/javassist/bytecode/AttributeInfo.java]: The original code needs at most two comparison but the new code needs four in the worst case.

**Comment by chibash** [comment]
As for redundant else statements, I'm glad to apply your fixes but I might personally prefer redundant else because it looks more functional.  :-)

File: src/main/javassist/Loader.java [hunk 2]
```diff
@@ -362,7 +362,7 @@ protected Class findClass(String name) throws ClassNotFoundException {
         int i = name.lastIndexOf('.');
         if (i != -1) {
             String pname = name.substring(0, i);
-            if (getPackage(pname) == null)
+            if (getDefinedPackage(pname) == null)
```
[the remark this answers is not in the document]
Comment (nickl-) [reply src/main/javassist/Loader.java]: Well spotted! I will see how this affects backward compatibility.

File: src/main/javassist/bytecode/AttributeInfo.java (hunk 1 above)
Comment (nickl-) [reply src/main/javassist/bytecode/AttributeInfo.java]: Yes this is correct, the change was applied to improve readability and reduce the so called "What the?!?!" factor. I believe there were 2 instances of this same pattern.

We do sacrifice an additional check (in the worst case) but gain ease of understanding.

**Comment by nickl-** [comment]
Thank you for applying this with the redundant else statement fixes, but I was more than prepared to revert any changes you did not agree with.

**Pull request merged**